File: 169-majority-element/majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        // using hashing
        // unordered_map<int, int> mpp;
        // for (int i = 0; i < nums.size(); i++) {
        //     mpp[nums[i]]++;
        // }

        // for (auto it : mpp) {
        //     if (it.second > (nums.size() / 2)) {
        //         return it.first;
        //     }
        // }
        // return -1;

        // Moore's Voting Algorithm
        int el, cnt = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (cnt == 0) {
                el = nums[i];
                cnt = 1;
            }

            else if (nums[i] == el) {
                cnt++;
            } else {
                cnt--;
            }
        }

        cnt = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] == el) {
                cnt++;
            }
        }
        if (cnt > nums.size() / 2) {
            return el;
        }
        return -1;
    }
};

File: 169-majority-element/majority-element.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        nums.sort()
        c=1
        curr=nums[0]
        if len(nums)==1:
            return curr
        for k in range(1,len(nums)):
            if curr==nums[k]:
                c+=1
            else:
                if c>(len(nums)//2):
                    return curr
                curr=nums[k]
                c=1
        if c>=(len(nums)//2):
                    return curr

        